Jenkinsfile

```diff
@@ -88,93 +88,237 @@ try {
+                                if (compiler == 'msvc') {
+                                    powershell "Remove-Item -Path \"${bldlabel}.txt\" -Force -ErrorAction Ignore"
+                                    // we capture stdout to variable because I could
+                                    // not figure out how to make powershell redirect internally
+                                    output = powershell (
+                                            returnStdout: true,
+                                            script: '''
+Invoke-BatchFile "${env:ProgramFiles(x86)}\\Microsoft Visual Studio\\2017\\Community\\VC\\Auxiliary\\Build\\vcvarsall.bat" x86_amd64
+$log="''' + "${bldlabel}.txt" + '''"
+Start-Transcript -Path $log
+Get-ChildItem env:* | Sort-Object name
+cl
+cmake --version
+New-Item -ItemType Directory -Force -Path "build/$env:BUILD_DIR" -ErrorAction Stop
+try {
+    Push-Location "build/$env:BUILD_DIR"
+    if ($env:NINJA_BUILD -eq "true") {
+        cmake -G"Ninja" -Dtarget="$env:COMPILER.$env:TARGET" -DCMAKE_VERBOSE_MAKEFILE=ON ../..
+    }
+    else {
+        cmake -G"Visual Studio 15 2017 Win64" -Dtarget="$env:COMPILER.$env:TARGET" -DCMAKE_VERBOSE_MAKEFILE=ON ../..
+    }
+    if ($LastExitCode -ne 0) { throw "CMake failed" }
+
+    ## as of 01/2018, DO NOT USE cmake to run the actual build step. for some
+    ## reason, cmake spawning the build under jenkins causes MSBUILD/ninja to
+    ## get stuck at the end of the build. Perhaps cmake is spawning
+    ## incorrectly or failing to pass certain params
+
+    if ($env:NINJA_BUILD -eq "true") {
+        ninja -j $env:NUMBER_OF_PROCESSORS -v
+    }
+    else {
+        msbuild /fl /m /nr:false /p:Configuration="$env:TARGET" /p:Platform=x64 /p:GenerateFullPaths=True /v:normal /nologo /clp:"ShowCommandLine;DisableConsoleColor" rippled.vcxproj
+    }
+    if ($LastExitCode -ne 0) { throw "CMake build failed" }
+
+    $exe = "./$env:TARGET/rippled"
+    if ($env:NINJA_BUILD -eq "true") {
+        $exe = "./rippled"
+    }
+    "Exe is at $exe"
+    $params = '--unittest', '--quiet', '--unittest-log'
+    if ($env:PARALLEL_TESTS -eq "true") {
+        $params = $params += "--unittest-jobs=$env:NUMBER_OF_PROCESSORS"
+    }
+    & $exe $params
+    if ($LastExitCode -ne 0) { throw "Unit tests failed" }
+}
+catch {
+    throw
+}
+finally {
+    Pop-Location
+    Stop-Transcript
+}
+''')
+                                    writeFile(
+                                        file: "${bldlabel}.txt",
+                                        text: output)
+                                }
+                                else {
+                                    sh "rm -fv ${bldlabel}.txt"
+                                    def bld_cmds = ""
+                                    if (bldtype == "docs") { // docs build
+                                        bld_cmds = '''
+cd docs
+rm -rf html_doc
+/usr/bin/time -p doxygen source.dox
+echo "0 tests total, 0 failures"
```

Comment:
How come this echo is here?

Reply:
it's a hack because my general passing criteria check requires no test failures (deduced by string matching the output file)...so the quick solution was just to echo this here and let the rest of the code do its thing. I could alternatively add some special case success check for the doc build if you think that's preferable.

Reply:
Not necessary. Maybe add a comment though.